[experiment, do not merge!] rewrite the DenseBitSet structure to only use 1 word on the stack

tage64 opened the pull request:
Modify DenseBitSet in the rustc_index crate so that it only uses one word on the stack instead of 4 words as before, allowing for faster clones. The downside is that it may at most store 63 elements on the stack as aposed to 128 for the previous implementation.

r? lqd

This is experimental so far and I mostly want a perf run to measure
the performance.

Commit:
Rewrite the DenseBitSet structure to only use one word on the stack.

This commit modifies DenseBitSet so that it only uses one word on the
stack instead of 4 words as before, allowing for faster clones. The
downside is that it may at most store 63 elements on the stack as aposed
to 128 for the previous implementation.

lqd commented:
(sorry for the unintended pings, everyone)

cc @nnethercote this is the bitset work I was telling you about. It did look interesting in my tests, ~= 0.5% wins on a lot of primary benchmarks.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15157161816/job/42614665016) [(plain)](https://github.com/rust-lang/rust/commit/3ff2c8852ecd9bfdf6a90606d16a7fcb6a3b0c45/checks/42614665016/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

thread 'rustc' panicked at /checkout/compiler/rustc_index/src/bit_set/dense_bit_set.rs:118:13:
attempt to shift right with overflow
stack backtrace:
   0:     0x7f2247d2b610 - <<std[d467a0f486057218]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[5399894434fbec0e]::fmt::Display>::fmt
   1:     0x7f2247d87e7f - core[5399894434fbec0e]::fmt::write
   2:     0x7f2247d1f1e9 - <std[d467a0f486057218]::sys::stdio::unix::Stderr as std[d467a0f486057218]::io::Write>::write_fmt
   3:     0x7f2247d2b4b2 - <std[d467a0f486057218]::sys::backtrace::BacktraceLock>::print
   4:     0x7f2247d2fbc8 - std[d467a0f486057218]::panicking::default_hook::{closure#0}
   5:     0x7f2247d2f962 - std[d467a0f486057218]::panicking::default_hook
   6:     0x7f22433e8714 - std[d467a0f486057218]::panicking::update_hook::<alloc[55daa3099f06b8bf]::boxed::Box<rustc_driver_impl[def0c2fbec255d7a]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7f2247d30743 - std[d467a0f486057218]::panicking::rust_panic_with_hook
   8:     0x7f2247d3032a - std[d467a0f486057218]::panicking::begin_panic_handler::{closure#0}
   9:     0x7f2247d2bc29 - std[d467a0f486057218]::sys::backtrace::__rust_end_short_backtrace::<std[d467a0f486057218]::panicking::begin_panic_handler::{closure#0}, !>
  10:     0x7f2247d2ff6d - __rustc[764efdf202547135]::rust_begin_unwind
  11:     0x7f2243394d80 - core[5399894434fbec0e]::panicking::panic_fmt
  12:     0x7f22433958b7 - core[5399894434fbec0e]::panicking::panic_const::panic_const_shr_overflow
  13:     0x7f22453f5c8c - rustc_mir_dataflow[690326a5ff6dbf69]::impls::storage_liveness::always_storage_live_locals
  14:     0x7f224520ceaa - <rustc_const_eval[9dd79d92b9ca9e69]::interpret::eval_context::InterpCx<rustc_const_eval[9dd79d92b9ca9e69]::const_eval::machine::CompileTimeMachine>>::storage_live_for_always_live_locals
  15:     0x7f22453a23cf - rustc_const_eval[9dd79d92b9ca9e69]::const_eval::eval_queries::eval_to_allocation_raw_provider
  16:     0x7f224610cbc0 - rustc_query_impl[9ceb55c5d78c945b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9ceb55c5d78c945b]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 24usize]>>
  17:     0x7f2246022226 - <rustc_query_impl[9ceb55c5d78c945b]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2} as core[5399894434fbec0e]::ops::function::FnOnce<(rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, rustc_middle[7a2b5353e52816d1]::ty::PseudoCanonicalInput<rustc_middle[7a2b5353e52816d1]::mir::interpret::GlobalId>)>>::call_once
  18:     0x7f2245f2af6b - rustc_query_system[e67bf3dc4a467b1e]::query::plumbing::try_execute_query::<rustc_query_impl[9ceb55c5d78c945b]::DynamicConfig<rustc_query_system[e67bf3dc4a467b1e]::query::caches::DefaultCache<rustc_middle[7a2b5353e52816d1]::ty::PseudoCanonicalInput<rustc_middle[7a2b5353e52816d1]::mir::interpret::GlobalId>, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[9ceb55c5d78c945b]::plumbing::QueryCtxt, false>
  19:     0x7f22463900bd - rustc_query_impl[9ceb55c5d78c945b]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  20:     0x7f22453739f4 - rustc_const_eval[9dd79d92b9ca9e69]::const_eval::valtrees::eval_to_valtree
  21:     0x7f22437a36eb - <rustc_const_eval[9dd79d92b9ca9e69]::provide::{closure#0} as core[5399894434fbec0e]::ops::function::FnOnce<(rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, rustc_middle[7a2b5353e52816d1]::ty::PseudoCanonicalInput<rustc_middle[7a2b5353e52816d1]::mir::interpret::GlobalId>)>>::call_once
  22:     0x7f22460fcc90 - rustc_query_impl[9ceb55c5d78c945b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9ceb55c5d78c945b]::query_impl::eval_to_valtree::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 24usize]>>
  23:     0x7f2245ffe026 - <rustc_query_impl[9ceb55c5d78c945b]::query_impl::eval_to_valtree::dynamic_query::{closure#2} as core[5399894434fbec0e]::ops::function::FnOnce<(rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, rustc_middle[7a2b5353e52816d1]::ty::PseudoCanonicalInput<rustc_middle[7a2b5353e52816d1]::mir::interpret::GlobalId>)>>::call_once
  24:     0x7f2245f2af6b - rustc_query_system[e67bf3dc4a467b1e]::query::plumbing::try_execute_query::<rustc_query_impl[9ceb55c5d78c945b]::DynamicConfig<rustc_query_system[e67bf3dc4a467b1e]::query::caches::DefaultCache<rustc_middle[7a2b5353e52816d1]::ty::PseudoCanonicalInput<rustc_middle[7a2b5353e52816d1]::mir::interpret::GlobalId>, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[9ceb55c5d78c945b]::plumbing::QueryCtxt, false>
  25:     0x7f22461d0f8d - rustc_query_impl[9ceb55c5d78c945b]::query_impl::eval_to_valtree::get_query_non_incr::__rust_end_short_backtrace
  26:     0x7f224747ec0c - <rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>::const_eval_global_id_for_typeck
  27:     0x7f224747d3f5 - <rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>::const_eval_resolve_for_typeck
  28:     0x7f22468fca01 - rustc_trait_selection[eaf972a72906f2cd]::traits::try_evaluate_const
  29:     0x7f22466c3806 - <rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate as rustc_next_trait_solver[b3e63bc3b229d9e0]::delegate::SolverDelegate>::evaluate_const
  30:     0x7f22469c870b - <rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::compute_normalizes_to_goal::{closure#0}
  31:     0x7f22469b05ea - <rustc_type_ir[ee13037aa9fea95d]::search_graph::SearchGraph<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate>, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_goal_in_task::<&mut <rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>
  32:     0x7f22469046d8 - <rustc_query_system[e67bf3dc4a467b1e]::dep_graph::graph::DepGraph<rustc_middle[7a2b5353e52816d1]::dep_graph::DepsType>>::with_anon_task::<rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, <rustc_type_ir[ee13037aa9fea95d]::search_graph::SearchGraph<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate>, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::with_new_goal<<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>::{closure#2}, (rustc_type_ir[ee13037aa9fea95d]::search_graph::StackEntry<rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>, core[5399894434fbec0e]::result::Result<rustc_type_ir[ee13037aa9fea95d]::canonical::Canonical<rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, rustc_type_ir[ee13037aa9fea95d]::solve::Response<rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>, rustc_type_ir[ee13037aa9fea95d]::solve::NoSolution>)>
  33:     0x7f22469ad9d9 - <rustc_type_ir[ee13037aa9fea95d]::search_graph::SearchGraph<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate>, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::with_new_goal::<<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>
  34:     0x7f22469cdfbb - <rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_goal_raw
  35:     0x7f22469cfd4a - <rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::try_evaluate_added_goals
  36:     0x7f22469b0d21 - <rustc_type_ir[ee13037aa9fea95d]::search_graph::SearchGraph<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate>, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_goal_in_task::<&mut <rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>
  37:     0x7f22469046d8 - <rustc_query_system[e67bf3dc4a467b1e]::dep_graph::graph::DepGraph<rustc_middle[7a2b5353e52816d1]::dep_graph::DepsType>>::with_anon_task::<rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, <rustc_type_ir[ee13037aa9fea95d]::search_graph::SearchGraph<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate>, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::with_new_goal<<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>::{closure#2}, (rustc_type_ir[ee13037aa9fea95d]::search_graph::StackEntry<rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>, core[5399894434fbec0e]::result::Result<rustc_type_ir[ee13037aa9fea95d]::canonical::Canonical<rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, rustc_type_ir[ee13037aa9fea95d]::solve::Response<rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>, rustc_type_ir[ee13037aa9fea95d]::solve::NoSolution>)>
  38:     0x7f22469ad9d9 - <rustc_type_ir[ee13037aa9fea95d]::search_graph::SearchGraph<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate>, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::with_new_goal::<<rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_canonical_goal::{closure#0}::{closure#0}::{closure#0}>
  39:     0x7f22469cdfbb - <rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_goal_raw
  40:     0x7f22469cd206 - <rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::evaluate_goal
  41:     0x7f22469a0eb5 - <rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate, rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>>::enter_root::<core[5399894434fbec0e]::result::Result<(rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::HasChanged, rustc_type_ir[ee13037aa9fea95d]::solve::Certainty), rustc_type_ir[ee13037aa9fea95d]::solve::NoSolution>, <rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate as rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::SolverDelegateEvalExt>::evaluate_root_goal::{closure#0}>
  42:     0x7f2246705f8e - <rustc_infer[ed8d7f599faefa29]::infer::InferCtxt as rustc_type_ir[ee13037aa9fea95d]::infer_ctxt::InferCtxtLike>::probe::<core[5399894434fbec0e]::result::Result<(rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::HasChanged, rustc_type_ir[ee13037aa9fea95d]::solve::Certainty), rustc_type_ir[ee13037aa9fea95d]::solve::NoSolution>, <rustc_trait_selection[eaf972a72906f2cd]::solve::delegate::SolverDelegate as rustc_next_trait_solver[b3e63bc3b229d9e0]::solve::eval_ctxt::SolverDelegateEvalExt>::root_goal_may_hold_with_depth::{closure#0}>
  43:     0x7f22468e3c56 - rustc_trait_selection[eaf972a72906f2cd]::traits::coherence::overlap
  44:     0x7f22468e2868 - rustc_trait_selection[eaf972a72906f2cd]::traits::coherence::overlapping_impls
  45:     0x7f2246a2dc4f - <rustc_middle[7a2b5353e52816d1]::traits::specialization_graph::Children as rustc_trait_selection[eaf972a72906f2cd]::traits::specialize::specialization_graph::ChildrenExt>::insert
  46:     0x7f2246a2f7b3 - <rustc_middle[7a2b5353e52816d1]::traits::specialization_graph::Graph as rustc_trait_selection[eaf972a72906f2cd]::traits::specialize::specialization_graph::GraphExt>::insert
  47:     0x7f224675d724 - rustc_trait_selection[eaf972a72906f2cd]::traits::specialize::specialization_graph_provider
  48:     0x7f224611129a - rustc_query_impl[9ceb55c5d78c945b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9ceb55c5d78c945b]::query_impl::specialization_graph_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 8usize]>>
  49:     0x7f224602bb59 - <rustc_query_impl[9ceb55c5d78c945b]::query_impl::specialization_graph_of::dynamic_query::{closure#2} as core[5399894434fbec0e]::ops::function::FnOnce<(rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, rustc_span[29c4001ea33351c8]::def_id::DefId)>>::call_once
  50:     0x7f2245f103e6 - rustc_query_system[e67bf3dc4a467b1e]::query::plumbing::try_execute_query::<rustc_query_impl[9ceb55c5d78c945b]::DynamicConfig<rustc_query_system[e67bf3dc4a467b1e]::query::caches::DefIdCache<rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[9ceb55c5d78c945b]::plumbing::QueryCtxt, false>
  51:     0x7f2246312cf8 - rustc_query_impl[9ceb55c5d78c945b]::query_impl::specialization_graph_of::get_query_non_incr::__rust_end_short_backtrace
  52:     0x7f22474ae0a3 - rustc_middle[7a2b5353e52816d1]::traits::specialization_graph::ancestors
  53:     0x7f22475eb54a - <rustc_middle[7a2b5353e52816d1]::ty::trait_def::TraitDef>::ancestors
  54:     0x7f224428788b - rustc_hir_analysis[65326a5424c475b7]::impl_wf_check::min_specialization::check_min_specialization
  55:     0x7f22440809b6 - rustc_hir_analysis[65326a5424c475b7]::impl_wf_check::check_impl_wf
  56:     0x7f224416cf62 - rustc_hir_analysis[65326a5424c475b7]::check::wfcheck::check_well_formed
  57:     0x7f2246101c05 - rustc_query_impl[9ceb55c5d78c945b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9ceb55c5d78c945b]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 1usize]>>
  58:     0x7f2246008fa5 - <rustc_query_impl[9ceb55c5d78c945b]::query_impl::check_well_formed::dynamic_query::{closure#2} as core[5399894434fbec0e]::ops::function::FnOnce<(rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, rustc_span[29c4001ea33351c8]::def_id::LocalDefId)>>::call_once
  59:     0x7f2245f8dfe2 - rustc_query_system[e67bf3dc4a467b1e]::query::plumbing::try_execute_query::<rustc_query_impl[9ceb55c5d78c945b]::DynamicConfig<rustc_data_structures[a1c7d19500cd477f]::vec_cache::VecCache<rustc_span[29c4001ea33351c8]::def_id::LocalDefId, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[e67bf3dc4a467b1e]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[9ceb55c5d78c945b]::plumbing::QueryCtxt, false>
  60:     0x7f22461a921c - rustc_query_impl[9ceb55c5d78c945b]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  61:     0x7f2244493356 - rustc_middle[7a2b5353e52816d1]::query::plumbing::query_ensure_error_guaranteed::<rustc_data_structures[a1c7d19500cd477f]::vec_cache::VecCache<rustc_span[29c4001ea33351c8]::def_id::LocalDefId, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[e67bf3dc4a467b1e]::dep_graph::graph::DepNodeIndex>, ()>
  62:     0x7f2244495248 - rustc_data_structures[a1c7d19500cd477f]::sync::parallel::try_par_for_each_in::<&[rustc_hir[1492663b1d663033]::hir::ItemId], rustc_span[29c4001ea33351c8]::ErrorGuaranteed, <rustc_middle[7a2b5353e52816d1]::hir::ModuleItems>::par_items<rustc_hir_analysis[65326a5424c475b7]::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>
  63:     0x7f2244181f81 - rustc_hir_analysis[65326a5424c475b7]::check::wfcheck::check_mod_type_wf
  64:     0x7f22461019f5 - rustc_query_impl[9ceb55c5d78c945b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9ceb55c5d78c945b]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 1usize]>>
  65:     0x7f2246008af5 - <rustc_query_impl[9ceb55c5d78c945b]::query_impl::check_mod_type_wf::dynamic_query::{closure#2} as core[5399894434fbec0e]::ops::function::FnOnce<(rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, rustc_span[29c4001ea33351c8]::def_id::LocalModDefId)>>::call_once
  66:     0x7f2245f4eb1b - rustc_query_system[e67bf3dc4a467b1e]::query::plumbing::try_execute_query::<rustc_query_impl[9ceb55c5d78c945b]::DynamicConfig<rustc_query_system[e67bf3dc4a467b1e]::query::caches::DefaultCache<rustc_span[29c4001ea33351c8]::def_id::LocalModDefId, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[9ceb55c5d78c945b]::plumbing::QueryCtxt, false>
  67:     0x7f22463d519c - rustc_query_impl[9ceb55c5d78c945b]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  68:     0x7f22444a7eef - <rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>::par_hir_for_each_module::<rustc_hir_analysis[65326a5424c475b7]::check_crate::{closure#0}::{closure#0}>::{closure#0}
  69:     0x7f224449361a - rustc_data_structures[a1c7d19500cd477f]::sync::parallel::par_for_each_in::<&rustc_hir[1492663b1d663033]::hir_id::OwnerId, &[rustc_hir[1492663b1d663033]::hir_id::OwnerId], <rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>::par_hir_for_each_module<rustc_hir_analysis[65326a5424c475b7]::check_crate::{closure#0}::{closure#0}>::{closure#0}>
  70:     0x7f224408495c - rustc_hir_analysis[65326a5424c475b7]::check_crate
  71:     0x7f22437b1b72 - rustc_interface[4fd74335141639b]::passes::analysis
  72:     0x7f224611c123 - rustc_query_impl[9ceb55c5d78c945b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9ceb55c5d78c945b]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 0usize]>>
  73:     0x7f2246044461 - <rustc_query_impl[9ceb55c5d78c945b]::query_impl::analysis::dynamic_query::{closure#2} as core[5399894434fbec0e]::ops::function::FnOnce<(rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt, ())>>::call_once
  74:     0x7f2245f13ea9 - rustc_query_system[e67bf3dc4a467b1e]::query::plumbing::try_execute_query::<rustc_query_impl[9ceb55c5d78c945b]::DynamicConfig<rustc_query_system[e67bf3dc4a467b1e]::query::caches::SingleCache<rustc_middle[7a2b5353e52816d1]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[9ceb55c5d78c945b]::plumbing::QueryCtxt, false>
  75:     0x7f22462b9ed2 - rustc_query_impl[9ceb55c5d78c945b]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  76:     0x7f2243402644 - <std[d467a0f486057218]::thread::local::LocalKey<core[5399894434fbec0e]::cell::Cell<*const ()>>>::with::<rustc_middle[7a2b5353e52816d1]::ty::context::tls::enter_context<<rustc_middle[7a2b5353e52816d1]::ty::context::GlobalCtxt>::enter<rustc_interface[4fd74335141639b]::passes::create_and_enter_global_ctxt<core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>, rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>>::{closure#1}, core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>>::{closure#0}, core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>>
  77:     0x7f22434ad032 - <rustc_middle[7a2b5353e52816d1]::ty::context::TyCtxt>::create_global_ctxt::<core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>, rustc_interface[4fd74335141639b]::passes::create_and_enter_global_ctxt<core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>, rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  78:     0x7f2243469e16 - <rustc_interface[4fd74335141639b]::passes::create_and_enter_global_ctxt<core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>, rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[5399894434fbec0e]::ops::function::FnOnce<(&rustc_session[e03ed62ea900b61e]::session::Session, rustc_middle[7a2b5353e52816d1]::ty::context::CurrentGcx, alloc[55daa3099f06b8bf]::sync::Arc<rustc_data_structures[a1c7d19500cd477f]::jobserver::Proxy>, &std[d467a0f486057218]::sync::once_lock::OnceLock<rustc_middle[7a2b5353e52816d1]::ty::context::GlobalCtxt>, &rustc_data_structures[a1c7d19500cd477f]::sync::worker_local::WorkerLocal<rustc_middle[7a2b5353e52816d1]::arena::Arena>, &rustc_data_structures[a1c7d19500cd477f]::sync::worker_local::WorkerLocal<rustc_hir[1492663b1d663033]::Arena>, rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  79:     0x7f22433f1756 - <alloc[55daa3099f06b8bf]::boxed::Box<dyn for<'a> core[5399894434fbec0e]::ops::function::FnOnce<(&'a rustc_session[e03ed62ea900b61e]::session::Session, rustc_middle[7a2b5353e52816d1]::ty::context::CurrentGcx, alloc[55daa3099f06b8bf]::sync::Arc<rustc_data_structures[a1c7d19500cd477f]::jobserver::Proxy>, &'a std[d467a0f486057218]::sync::once_lock::OnceLock<rustc_middle[7a2b5353e52816d1]::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures[a1c7d19500cd477f]::sync::worker_local::WorkerLocal<rustc_middle[7a2b5353e52816d1]::arena::Arena<'a>>, &'a rustc_data_structures[a1c7d19500cd477f]::sync::worker_local::WorkerLocal<rustc_hir[1492663b1d663033]::Arena<'a>>, rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}::{closure#2}), Output = core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>>> as core[5399894434fbec0e]::ops::function::FnOnce<(&rustc_session[e03ed62ea900b61e]::session::Session, rustc_middle[7a2b5353e52816d1]::ty::context::CurrentGcx, alloc[55daa3099f06b8bf]::sync::Arc<rustc_data_structures[a1c7d19500cd477f]::jobserver::Proxy>, &std[d467a0f486057218]::sync::once_lock::OnceLock<rustc_middle[7a2b5353e52816d1]::ty::context::GlobalCtxt>, &rustc_data_structures[a1c7d19500cd477f]::sync::worker_local::WorkerLocal<rustc_middle[7a2b5353e52816d1]::arena::Arena>, &rustc_data_structures[a1c7d19500cd477f]::sync::worker_local::WorkerLocal<rustc_hir[1492663b1d663033]::Arena>, rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}::{closure#2})>>::call_once
  80:     0x7f22434680e7 - rustc_interface[4fd74335141639b]::passes::create_and_enter_global_ctxt::<core[5399894434fbec0e]::option::Option<rustc_interface[4fd74335141639b]::queries::Linker>, rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}::{closure#2}>
  81:     0x7f2243497f69 - rustc_span[29c4001ea33351c8]::create_session_globals_then::<(), rustc_interface[4fd74335141639b]::util::run_in_thread_with_globals<rustc_interface[4fd74335141639b]::util::run_in_thread_pool_with_globals<rustc_interface[4fd74335141639b]::interface::run_compiler<(), rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
  82:     0x7f224341f8a9 - std[d467a0f486057218]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[4fd74335141639b]::util::run_in_thread_with_globals<rustc_interface[4fd74335141639b]::util::run_in_thread_pool_with_globals<rustc_interface[4fd74335141639b]::interface::run_compiler<(), rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  83:     0x7f22434225a4 - <<std[d467a0f486057218]::thread::Builder>::spawn_unchecked_<rustc_interface[4fd74335141639b]::util::run_in_thread_with_globals<rustc_interface[4fd74335141639b]::util::run_in_thread_pool_with_globals<rustc_interface[4fd74335141639b]::interface::run_compiler<(), rustc_driver_impl[def0c2fbec255d7a]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[5399894434fbec0e]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  84:     0x7f2247d35a22 - <std[d467a0f486057218]::sys::pal::unix::thread::Thread>::new::thread_start
  85:     0x7f224266bac3 - <unknown>
  86:     0x7f22426fd850 - <unknown>
  87:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.
---
warning: the ICE couldn't be written to `/checkout/rustc-ice-2025-05-21T08_30_04-12635.txt`: Read-only file system (os error 30)

note: rustc 1.89.0-nightly (4b3ec1d2b 2025-05-21) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debug-assertions=on -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C split-debuginfo=off -C prefer-dynamic -Z inline-mir -Z inline-mir-preserve-debug -Z mir_strip_debuginfo=locals-in-tiny-functions -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C embed-bitcode=yes -C force-frame-pointers=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [eval_to_allocation_raw] const-evaluating + checking `array::<impl at library/core/src/array/mod.rs:484:9: 484:36>::{constant#0}`
#1 [eval_to_valtree] evaluating type-level constant
... and 4 other queries... use `env RUST_BACKTRACE=1` to see the full query stack
[RUSTC-TIMING] core test:false 7.794
error: could not compile `core` (lib)

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc /checkout/obj/build/bootstrap/debug/rustc --crate-name core --edition=2024 library/core/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 --warn=unexpected_cfgs --check-cfg 'cfg(bootstrap)' --check-cfg 'cfg(no_fp_fmt_parse)' --check-cfg 'cfg(feature, values(any()))' --check-cfg 'cfg(target_has_reliable_f16)' --check-cfg 'cfg(target_has_reliable_f16_math)' --check-cfg 'cfg(target_has_reliable_f128)' --check-cfg 'cfg(target_has_reliable_f128_math)' -C debug-assertions=on --check-cfg 'cfg(docsrs,test)' --check-cfg 'cfg(feature, values("debug_refcell", "debug_typeid", "optimize_for_size", "panic_immediate_abort"))' -C metadata=4f9217aff00f2171 -C extra-filename=-cb7d263a0f2eedd1 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps -Csymbol-mangling-version=v0 '--check-cfg=cfg(feature,values(any()))' -Zunstable-options '--check-cfg=cfg(bootstrap)' -Zmacro-backtrace -Csplit-debuginfo=off -Cprefer-dynamic -Zinline-mir -Zinline-mir-preserve-debug -Zmir_strip_debuginfo=locals-in-tiny-functions -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Alinker-messages -Cembed-bitcode=yes -Cforce-frame-pointers=yes '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' -Z binary-dep-depinfo` (exit status: 101)
Build completed unsuccessfully in 0:05:11
  local time: Wed May 21 08:30:11 UTC 2025
  network time: Wed, 21 May 2025 08:30:12 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:hourglass: Trying commit 3ff2c8852ecd9bfdf6a90606d16a7fcb6a3b0c45 with merge 693cccaceb81d959dbf823da027464c8655b8b57...
<!-- homu: {"type":"TryBuildStarted","head_sha":"3ff2c8852ecd9bfdf6a90606d16a7fcb6a3b0c45","merge_sha":"693cccaceb81d959dbf823da027464c8655b8b57"} -->

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

nnethercote commented:
Looks interesting. Unfortunately being a single commit makes it incredibly hard to review, especially with so much code being moved into `dense_bit_set.rs`. If it gets beyond the draft stage I think the commits need to be structured differently to make review easier. E.g. the code doesn't need to move into `dense_bit_set.rs`, or if that's really desired, it should be moved in a first commit and then modified in a second commit. And any other changes are made that are independent (e.g. changing `MixedBitSet::Small` to `Self::Small`) can be done in separate commits as well.

lqd commented:
Yeah this is just for the perf run I think. The branch I was looking at (now unfortunately gone) had many smaller commits that were easier to review, eg the last 20 of https://github.com/rust-lang/rust/compare/master...tage64:rust:replace_predecessors

tage64 commented:
I squashed all commits into one to make rebasing on upstream easier. Maybe that was not necessary but as you said, this is just for experiments. The original commit history is still there in [tage64/rust@thin_bit_set_old](github.com/tage64/rust/tree/thin_bit_set_old).

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/15157385609/job/42623672301)
Build commit: 693cccaceb81d959dbf823da027464c8655b8b57 (`693cccaceb81d959dbf823da027464c8655b8b57`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/15157385609/job/42623672301"},"merge_sha":"693cccaceb81d959dbf823da027464c8655b8b57"} -->

rust-timer commented:
Queued 693cccaceb81d959dbf823da027464c8655b8b57 with parent 87b454156998b945cf161c951f0fbc20ac292cf6, future [comparison URL](https://perf.rust-lang.org/compare.html?start=87b454156998b945cf161c951f0fbc20ac292cf6&end=693cccaceb81d959dbf823da027464c8655b8b57).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.8 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

Commit:
DenseBitSet: Solve a few bugs related to overflow in SHR.

rust-timer commented:
Finished benchmarking commit (693cccaceb81d959dbf823da027464c8655b8b57): [comparison URL](https://perf.rust-lang.org/compare.html?start=87b454156998b945cf161c951f0fbc20ac292cf6&end=693cccaceb81d959dbf823da027464c8655b8b57&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=87b454156998b945cf161c951f0fbc20ac292cf6&end=693cccaceb81d959dbf823da027464c8655b8b57&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-1.1%, -0.1%] | 83    |
| Improvements ✅ <br /> (secondary) | -1.0% | [-3.3%, -0.1%] | 50    |
| All ❌✅ (primary)                 | -0.5% | [-1.1%, -0.1%] | 83    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=87b454156998b945cf161c951f0fbc20ac292cf6&end=693cccaceb81d959dbf823da027464c8655b8b57&stat=max-rss)
<details>
<summary>Results (primary -1.1%, secondary -1.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.8%, 0.9%]   | 2     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.1%, -1.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-5.2%, -0.4%] | 4     |
| All ❌✅ (primary)                 | -1.1% | [-1.1%, -1.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=87b454156998b945cf161c951f0fbc20ac292cf6&end=693cccaceb81d959dbf823da027464c8655b8b57&stat=cycles:u)
<details>
<summary>Results (secondary -0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.8%  | [1.5%, 2.1%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-2.0%, -0.7%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=87b454156998b945cf161c951f0fbc20ac292cf6&end=693cccaceb81d959dbf823da027464c8655b8b57&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0% | [0.0%, 0.1%] | 48    |
| Regressions ❌ <br /> (secondary)  | 0.1% | [0.0%, 0.1%] | 25    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.0% | [0.0%, 0.1%] | 48    |
</details>

**Bootstrap**: 775.275s -> 774.785s (-0.06%)
**Artifact size**: 365.51 MiB -> 365.51 MiB (0.00%)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15161907041/job/42629864713) [(plain)](https://github.com/rust-lang/rust/commit/eddd1cc8610e0d15808fea98c021080242ae87d5/checks/42629864713/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting rustc_index v0.0.0 (/checkout/compiler/rustc_index)
error: unresolved link to `OnceCell`
  --> compiler/rustc_index/src/bit_set/dense_bit_set.rs:29:12
   |
29 | /// in a [`OnceCell`] or an [`Option`]—you can simply create an empty set and populate it if needed.
   |            ^^^^^^^^ no item named `OnceCell` in scope
   |
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: unresolved link to `MixedBitSet`
  --> compiler/rustc_index/src/bit_set/dense_bit_set.rs:32:73
   |
32 | /// long runs of set or unset bits), it may be more efficient to use a [MixedBitSet] or an
   |                                                                         ^^^^^^^^^^^ no item named `MixedBitSet` in scope
   |
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: could not document `rustc_index`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:44
  local time: Wed May 21 12:44:07 UTC 2025
  network time: Wed, 21 May 2025 12:44:08 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Commit:
fix documentation errors

Commit:
remove a test checking that the hash for bit sets of different domain
sizes are different

The new implementation of DenseBitSet doesn't store the exact domain
size, so of course the hash values for identical sets with different
domain sizes may be equal.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15165694031/job/42642881593) [(plain)](https://github.com/rust-lang/rust/commit/f17b8f7d566695149b42878c9f8bab4a3c20bfca/checks/42642881593/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 20.5s done
#18 DONE 26.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] proc_macro test:false 5.414
[RUSTC-TIMING] test test:false 9.162
    Finished `release` profile [optimized] target(s) in 1m 16s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 76.290
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.728
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.763
    Finished `release` profile [optimized] target(s) in 3.94s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.192
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.171
---
---- [run-make] tests/run-make/rustc-crates-on-stable stdout ----

error: rmake recipe failed to complete
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustc-crates-on-stable/rmake_out" && env -u RUSTFLAGS AR="ar" BUILD_ROOT="/checkout/obj/build/x86_64-unknown-linux-gnu" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RUSTC_DYLIB_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-19/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegendata codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-19/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_EXE_DYLIB_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustc-crates-on-stable/rmake"
stdout: none
--- stderr -------------------------------
command failed at line 41
Command { cmd: LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustc-crates-on-stable/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_STAGE="0" RUSTFLAGS="-Zallow-features=" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" "build" "--manifest-path" "/checkout/Cargo.toml" "--no-default-features" "--target-dir" "target" "-p" "rustc_type_ir" "-p" "rustc_next_trait_solver" "-p" "rustc_pattern_analysis" "-p" "rustc_lexer" "-p" "rustc_abi" "-p" "rustc_parse_format" "-p" "rustc_hashes", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo", defused: true, armed_location: Location { file: "src/tools/run-make-support/src/external_deps/cargo.rs", line: 8, col: 19 } }, already_executed: true }
output status: `exit status: 101`
=== STDOUT ===



=== STDERR ===
   Compiling proc-macro2 v1.0.95
```

</details>

Commit:
in rustc_index: fix compilation error by changing the rustc_serialize dependency to be required instead of optional

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15182354220/job/42694565358) [(plain)](https://github.com/rust-lang/rust/commit/44f13f44abd0df88cf82aa166f59c9a3c46cd384/checks/42694565358/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:637d52fd4b642580188810b2d213c11aa3a3d2219df89616a68115a83d72b2ef61353e7514d28184e1c711c4ba72be220d92f8a60438c5cd694caadb1fb646a3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] proc_macro test:false 5.683
[RUSTC-TIMING] test test:false 9.511
    Finished `release` profile [optimized] target(s) in 1m 19s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 79.889
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.773
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.959
    Finished `release` profile [optimized] target(s) in 4.02s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.206
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.179
---
---- [run-make] tests/run-make/rustc-crates-on-stable stdout ----

error: rmake recipe failed to complete
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustc-crates-on-stable/rmake_out" && env -u RUSTFLAGS AR="ar" BUILD_ROOT="/checkout/obj/build/x86_64-unknown-linux-gnu" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RUSTC_DYLIB_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-19/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegendata codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-19/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_EXE_DYLIB_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustc-crates-on-stable/rmake"
stdout: none
--- stderr -------------------------------
command failed at line 41
Command { cmd: LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustc-crates-on-stable/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_STAGE="0" RUSTFLAGS="-Zallow-features=" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" "build" "--manifest-path" "/checkout/Cargo.toml" "--no-default-features" "--target-dir" "target" "-p" "rustc_type_ir" "-p" "rustc_next_trait_solver" "-p" "rustc_pattern_analysis" "-p" "rustc_lexer" "-p" "rustc_abi" "-p" "rustc_parse_format" "-p" "rustc_hashes", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo", defused: true, armed_location: Location { file: "src/tools/run-make-support/src/external_deps/cargo.rs", line: 8, col: 19 } }, already_executed: true }
output status: `exit status: 101`
=== STDOUT ===



=== STDERR ===
   Compiling proc-macro2 v1.0.95
---
   Compiling unicode-xid v0.2.6
error[E0725]: the feature `dropck_eyepatch` is not in the list of allowed features
##[error]  --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/smallvec-1.15.0/src/lib.rs:96:45
   |
96 | #![cfg_attr(feature = "may_dangle", feature(dropck_eyepatch))]
   |                                             ^^^^^^^^^^^^^^^

   Compiling memchr v2.7.4
error[E0658]: `may_dangle` has unstable semantics and may be removed in the future
##[error]    --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/smallvec-1.15.0/src/lib.rs:2131:13
     |
2131 | unsafe impl<#[may_dangle] A: Array> Drop for SmallVec<A> {
     |             ^^^^^^^^^^^^^
     |
     = note: see issue #34761 <https://github.com/rust-lang/rust/issues/34761> for more information
     = help: add `#![feature(dropck_eyepatch)]` to the crate attributes to enable
     = note: this compiler was built on 2025-05-22; consider upgrading it if it is out of date

   Compiling unicode-properties v0.1.3
   Compiling rustc-hash v1.1.0
   Compiling rustc-hash v2.1.1
```

</details>

Commit:
in rustc_index: revert "fix compilation error by changing the rustc_serialize dependency to be required instead of optional", and introduce conditional compilation instead.

lqd commented:
@rustbot author
